python/ray/util/client/dataclient.py

```diff
@@ -101,54 +126,109 @@ def _process_response(self, response: Any) -> None:
-            # in practice we may need to think about the correct semantics
-            # here.
-            logger.info("Server disconnected from data channel")
-        else:
-            logger.exception("Got Error from data channel -- shutting down:")
-
-    def _shutdown(self, e: grpc.RpcError) -> None:
+        if not self.client_worker._can_reconnect(e):
+            logger.info("Unrecoverable error in data channel.")
+            logger.debug(e)
+            return False
+        logger.debug("Recoverable error in data channel.")
+        logger.debug(e)
+        return True
+
+    def _shutdown(self) -> None:
         """
         Shutdown the data channel
         """
         with self.lock:
             self._in_shutdown = True
-            self._last_exception = e
             self.cv.notify_all()
 
             callbacks = self.asyncio_waiting_data.values()
             self.asyncio_waiting_data = {}
 
-        # Abort async requests with the error.
-        err = ConnectionError("Failed during this or a previous request. "
-                              f"Exception that broke the connection: {e}")
+        if self._last_exception:
+            # Abort async requests with the error.
+            err = ConnectionError(
+                "Failed during this or a previous request. Exception that "
+                f"broke the connection: {self._last_exception}")
+        else:
+            err = ConnectionError(
+                "Request cannot be fulfilled because the data client has "
+                "disconnected.")
         for callback in callbacks:
-            callback(err)
+            if callback:
+                callback(err)
         # Since self._in_shutdown is set to True, no new item
         # will be added to self.asyncio_waiting_data
 
+    def _acknowledge(self, req_id: int) -> None:
+        """
+        Puts an acknowledge request on the request queue periodically.
+        Lock should be held before calling this. Used when an async or
+        blocking response is received.
+        """
+        assert self.lock.locked()
+        self._acknowledge_counter += 1
+        if self._acknowledge_counter % ACKNOWLEDGE_BATCH_SIZE == 0:
+            self.request_queue.put(
+                ray_client_pb2.DataRequest(
+                    acknowledge=ray_client_pb2.AcknowledgeRequest(
+                        req_id=req_id)))
+
+    def _reconnect_channel(self) -> None:
+        """
+        Attempts to reconnect the gRPC channel and resend outstanding
+        requests. First, the server is pinged to see if the current channel
+        still works. If the ping fails, then the current channel is closed
+        and replaced with a new one.
+
+        Once a working channel is available, a new request queue is made
+        and filled with any outstanding requests to be resent to the server.
+        """
+        # Ping the server to see if the current channel is reuseable
+        try:
```

Comment:
Can you please add details as you described to clarify that this is used for internal grcp reconnection...